crates/pessimistic-proof-core/src/aggchain_proof.rs

```diff
@@ -0,0 +1,103 @@
+pub enum AggchainType {
+    ECDSA = 0,
+    SP1 = 1,
+}
+
+pub type Vkey = [u32; 8];
+pub type StarkVkey = Vec<u8>;
+pub type StarkProof = Vec<u8>;
+
+/// Aggchain Proof which is either one ECDSA signature, or one stark proof.
+/// Contains all the necessary data for verification.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum AggchainProofData {
+    /// ECDSA signature.
+    ECDSA(AggchainProofECDSAData),
+    /// STARK proof generated with SP1 and its metadata.
+    SP1(AggchainProofSP1Data),
+}
+
+impl AggchainProofData {
+    /// Returns the aggchain hash
+    pub fn aggchain_hash(&self) -> Digest {
+        match &self {
+            AggchainProofData::ECDSA(aggchain_proof_ecdsa) => keccak256_combine([
+                &(AggchainType::ECDSA as u32).to_be_bytes(),
+                aggchain_proof_ecdsa.signer.as_slice(),
+            ]),
+            AggchainProofData::SP1(aggchain_proof_sp1) => keccak256_combine([
+                &(AggchainType::SP1 as u32).to_be_bytes(),
+                words_to_bytes_le(&aggchain_proof_sp1.aggchain_vkey).as_slice(),
+                aggchain_proof_sp1.aggchain_params.as_slice(),
+            ]),
+        }
+    }
+}
+
+/// ECDSA variant of the aggchain proof.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AggchainProofECDSAData {
+    /// Signer committing to the state transition.
+    pub signer: Address,
+    /// Signature committing to the state transition.
+    pub signature: Signature,
+}
+
+/// SP1 variant of the aggchain proof.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AggchainProofSP1Data {
+    /// Chain-specific commitment forwarded by the PP.
+    pub aggchain_params: Digest,
+    /// SP1 verifying key for the SP1 aggchain proof program.
+    pub aggchain_vkey: Vkey,
+    /// STARK proof.
+    pub stark_proof: StarkProof,
+}
+
+/// Public values to verify the SP1 aggchain proof.
+#[derive(Serialize, Deserialize)]
+pub struct AggchainProofPublicValues {
+    /// Previous local exit root.
+    pub prev_local_exit_root: Digest,
+    /// New local exit root.
+    pub new_local_exit_root: Digest,
+    /// L1 info root used to import bridge exits.
+    pub l1_info_root: Digest,
+    /// Origin network for which the proof was generated.
+    pub origin_network: u32,
+    /// Commitment to the imported bridge exits indexes.
+    pub commit_imported_bridge_exits: Digest,
+    /// Chain-specific commitment forwarded by the PP.
+    pub aggchain_params: Digest,
+}
+
+impl AggchainProofPublicValues {
+    pub fn hash(&self) -> [u8; 32] {
+        let public_values = [
+            self.prev_local_exit_root.as_slice(),
+            self.new_local_exit_root.as_slice(),
+            self.l1_info_root.as_slice(),
+            &self.origin_network.to_be_bytes(),
+            self.commit_imported_bridge_exits.as_slice(),
+            self.aggchain_params.as_slice(),
+        ]
+        .concat();
+
+        Sha256::digest(&public_values).into()
```

Comment:
Just for my understanding, why is it `sha256` and not `keccak` here?

Reply:
keccak is mainly used for the hashes which need to be recomputed on the solidity side, but here this hash is solely for the stark verification

Reply:
>Unresolving the conversation: I’d argue for using keccak still here

ah I don't think we can really choose here, the public values are expected to be hashed with sha256 I believe https://github.com/succinctlabs/sp1/blob/dev/crates/primitives/src/io.rs#L55-L79
https://github.com/succinctlabs/sp1/blob/dev/examples/aggregation/program/src/main.rs#L54-L55

>Wondering why is sha256 required / preferable in this case?

I'm not expert tbh, @Nashtare would know more

Reply:
From a security standpoint, SHA3 (keccak) is technically subtly superior in some regards to SHA2 (SHA256), but that doesn't matter here.
I agree the point of @Ekleog-Polygon of unifying where possible / meaningful the hash function used (unless there are counter-reasons) makes sense, however from a performance standpoint, I'd expect SHA2 circuit to be more efficient than KECCAK, which may impair performance of the consensus proof. Maybe worth benchmarking the cycle count.